misk/src/main/kotlin/misk/web/actions/ReadinessCheckAction.kt

```diff
@@ -7,25 +7,35 @@ import misk.web.AvailableWhenDegraded
 import misk.web.Get
 import misk.web.Response
 import misk.web.ResponseContentType
+import misk.web.WebConfig
 import misk.web.mediatype.MediaTypes
 import wisp.logging.getLogger
 import javax.inject.Inject
 import javax.inject.Provider
 import javax.inject.Singleton
+import kotlinx.coroutines.*
+import misk.healthchecks.HealthStatus
+import java.time.Clock
+import java.time.Instant
 
 private val logger = getLogger<ReadinessCheckAction>()
 
 @Singleton
 class ReadinessCheckAction @Inject internal constructor(
   private val serviceManagerProvider: Provider<ServiceManager>,
-  @JvmSuppressWildcards private val healthChecks: List<HealthCheck>
+  @JvmSuppressWildcards private val healthChecks: List<HealthCheck>,
+  val config: WebConfig,
+  val clock: Clock
 ) : WebAction {
 
+  var lastUpdate: Instant? = null
+  var statuses: List<HealthStatus> = listOf()
+
   @Get("/_readiness")
   @ResponseContentType(MediaTypes.APPLICATION_JSON)
   @Unauthenticated
   @AvailableWhenDegraded
-  fun readinessCheck(): Response<String> {
+  fun readinessCheck(): Response<String> = runBlocking {
```

Comment:
Why are we using coroutines? Misk/Jetty is blocking.

Reply:
I want the actual refresh to happen in the background, not on the request thread. Is there a better way to do this?

Reply:
Related to the executor pool question above, I think ideally we'd inject the thread pool misk [uses here](https://github.com/cashapp/misk/blob/78a95ce4e3f0a90fbdbe6d05486adcc96ff13145/misk/src/main/kotlin/misk/web/jetty/JettyService.kt#L70
https://github.com/tgregory-block/misk/blob/78a95ce4e3f0a90fbdbe6d05486adcc96ff13145/misk/src/main/kotlin/misk/web/jetty/JettyService.kt#L70) and then use its `submit` method. Would probably need to install the pool as an executor service module to make it injectable

Reply:
So first, I'm still learning how threads and coroutines work in Kotlin, so I'm happy to be corrected on any of this.

Reading the [coroutine docs](https://kotlinlang.org/docs/coroutine-context-and-dispatchers.html#dispatchers-and-threads) it sounds like the default dispatcher for a job is the parent's context. I think that means that by leaving the context blank, e.g. `launch(Job())` that the `healthExecutor` will be used, since its the context of the parent.

Does that sound right to you?

Reply:
I'm also not super familiar with kotlin coroutines (more familiar with Swift's oddly enough, but Kotlin coroutines are also designed around structured concurrency so it looks fairly similar), but my reading of the docs is that:
* `launch` with no specified `Dispatcher` will run in the context of its current coroutine scope
* Here, that scope comes from the `runBlocking` that's added in the PR
* `runBlocking` uses "an event loop on the current thread" by default

So indeed, it sounds like it will run on the thread the action already uses from `healthExecutor`. I'm not sure if there's any reason to be concerned about using `launch` in combination with a thread pool executor (e.g. any weirdness between the coroutine's "event loop" and the thread pool's own synchronous queue for work management?) @adrw are you able to shed any light on this?

Reply:
I have a couple reservations about extracting the `healthExecutor`.

First, I'm not sure that using a thread pool with 4-8 threads is what we want. If concurrent requests to readiness are made I think we want at most 1 thread refreshing the status. I think we want a dedicated thread (or single thread pool) for that.

Second, the complexity impact is awkward.

* `healthExecutor` is only created when there is a `health_port` mapped. From a DI perspective this means creating an optional binding and moving that logic out of `JettyService` into another module, then handling that nullability in both `JettyService` and `ReadinessCheckAction`.
  * In `ReadinessCheckAction` its not actually nullable, but will still be getting a nullable value.
* It exposes that executor for use in other places, which is almost certainly unwanted.

I would prefer to create a private executor inside of `ReadinessCheckAction`. @tgregory-block thoughts?

Reply:
Ah that's a good idea. It keeps one thread for running the checks themselves and leaves `healthExecutor` to handle the web requests, and also gets rid of the concurrency issue.